engine/src/main/java/org/pentaho/di/trans/steps/script/ScriptAddedFunctions.java

```diff
@@ -0,0 +1,2532 @@
+        cal.setTime( dArg1 );
+        return new Double( cal.get( Calendar.MONTH ) );
+      } else {
+        throw new RuntimeException( "The function call month requires 1 argument." );
+      }
+    } catch ( Exception e ) {
+      throw new RuntimeException( e.toString() );
+    }
+
+  }
+
+  public static Object quarter( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+    try {
+      if ( ArgList.length == 1 ) {
+        if ( isNull( ArgList[0] ) ) {
+          return new Double( Double.NaN );
+        } else if ( isUndefined( ArgList[0] ) ) {
+          return undefinedValue;
+        }
+        java.util.Date dArg1 = (java.util.Date) ArgList[0];
+        Calendar cal = Calendar.getInstance();
+        cal.setTime( dArg1 );
+
+        // Patch by Ingo Klose: calendar months start at 0 in java.
+        int iMonth = cal.get( Calendar.MONTH );
+        if ( iMonth <= 2 ) {
+          return new Double( 1 );
+        } else if ( iMonth <= 5 ) {
+          return new Double( 2 );
+        } else if ( iMonth <= 8 ) {
+          return new Double( 3 );
+        } else {
+          return new Double( 4 );
+        }
+      } else {
+        throw new RuntimeException( "The function call quarter requires 1 argument." );
+      }
+    } catch ( Exception e ) {
+      throw new RuntimeException( e.toString() );
+    }
+  }
+
+  public static Object week( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+    try {
+      if ( ArgList.length == 1 ) {
+        if ( isNull( ArgList[0] ) ) {
+          return new Double( Double.NaN );
+        } else if ( isUndefined( ArgList[0] ) ) {
+          return undefinedValue;
+        }
+        java.util.Date dArg1 = (java.util.Date) ArgList[0];
+        Calendar cal = Calendar.getInstance();
+        cal.setTime( dArg1 );
+        return new Double( cal.get( Calendar.WEEK_OF_YEAR ) );
+      } else {
+        throw new RuntimeException( "The function call week requires 1 argument." );
+      }
+    } catch ( Exception e ) {
+      throw new RuntimeException( e.toString() );
+    }
+  }
+
+  public static Object str2RegExp( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+    String[] strArr = null;
+    if ( ArgList.length == 2 ) {
+      try {
+        if ( isNull( ArgList, new int[] { 0, 1 } ) ) {
+          return null;
+        } else if ( isUndefined( ArgList, new int[] { 0, 1 } ) ) {
+          return undefinedValue;
+        }
+        String strToMatch = (String) ArgList[0];
+        Pattern p = Pattern.compile( (String) ArgList[1] );
+        Matcher m = p.matcher( strToMatch );
+        if ( m.matches() && m.groupCount() > 0 ) {
+          strArr = new String[m.groupCount()];
+          for ( int i = 1; i <= m.groupCount(); i++ ) {
+            strArr[i - 1] = m.group( i );
+          }
+        }
+      } catch ( Exception e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)